library(ggplot2)
library(TissueEnrich)
library(openxlsx)
library(RColorBrewer)
library(ggplotify)
library(pheatmap)
set.seed(0822)
setwd('/media/dell/0E54E2B554E29EA9/HanRunpeng/mregDC_project/Lab_Sequencing_Data/Code deposit/TRA expression analysis')
# DC2hm DEGs and TRA list -------------------------------------------------

  # Use LFC>=2 as a threshold as TRAs are usually lowly-expressed to ensure reliability
dc2hm.vs.cdc2.de.results <- read.xlsx('../DC2hm vs Immunogenic activated cDC2s/Outputs/tables/DC2hm_vs_cDC2s.xlsx',
                                      sheet = 'DE results', rowNames = T)
dc2hm.up.2 <- rownames(dc2hm.vs.cdc2.de.results[dc2hm.vs.cdc2.de.results$logFC>2 &
                                                  dc2hm.vs.cdc2.de.results$adj.P.Val < 0.05, ])
dc2hm.down.2 <- rownames(dc2hm.vs.cdc2.de.results[dc2hm.vs.cdc2.de.results$logFC <(-2) &
                                                    dc2hm.vs.cdc2.de.results$adj.P.Val < 0.05, ])
  # The aire-TRA list is generated by Sansom et al. (Genome Res. 2014)
aire.induced.tra <- read.csv('AIRE-regulating TRA.csv',
                             row.names = 1)
  # The cutoff is according to the orignial work: Sansom et al. (Genome Res. 2014)
aire.induced.tra <- toupper(aire.induced.tra[aire.induced.tra$Fold.Change < .5 & aire.induced.tra$FDR..BH.<0.05,
                                             1])


# TRAs in DC2hm -----------------------------------------------------------
# Plot the  Heatmap
merged.tpm <- read.csv('../Analysis on sorted cDCs/cDC_adjusted_TPM.csv',
                       row.names = 1, check.names = F)
merged.tpm.aire.tra <- merged.tpm[intersect(c(dc2hm.down.2, dc2hm.up.2),
                                            aire.induced.tra),5:12]
as.ggplot(pheatmap(merged.tpm.aire.tra, cluster_cols = F, cluster_rows = T,
                   scale = 'row', show_rownames = F, cellwidth = 10,
                   cellheight = 1))
ggsave('Outputs/figures/AIRE_TRA_Heatmap_up.pdf',
       width = 6, height = 6)

# TissueEnrich ------------------------------------------------------------

# Upregulated Tissue-enrich
gs <- GeneSet(
  geneIds = intersect(dc2hm.up.2,
                      aire.induced.tra),
  organism = "Homo Sapiens",                 # Specify the organism
  geneIdType = SymbolIdentifier(),           # Define gene ID type as Symbol
  setName = "DC2hmGenes"               # Optional: Give a name to the gene set
)
dc2hm.tissue.enrich.2 <- teEnrichment(gs)
# Fetch all hits
dc2hm.enrichment.hits.2 <- c()
for (i in names(dc2hm.tissue.enrich.2[[2]])) {
  dc2hm.enrichment.hits.2 <- c(dc2hm.enrichment.hits.2,
                               rownames(dc2hm.tissue.enrich.2[[2]][[i]]))
}
# Cross ref with another datasets
seEnrichmentOutput <- dc2hm.tissue.enrich.2[[1]]
enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput),
                                      row.names = rowData(seEnrichmentOutput)[,1]),
                           colData(seEnrichmentOutput)[,1])
enrichmentOutput$Tissue<-row.names(enrichmentOutput)
accent_palette <- colorRampPalette(brewer.pal(8, "Accent"))(35)
set2_palette <- colorRampPalette(brewer.pal(8, "Set2"))(35)
# Plotting the P-Values
ggplot(enrichmentOutput,aes(x=reorder(Tissue,Tissue.Specific.Genes),
                            y=Tissue.Specific.Genes,
                            label = Tissue.Specific.Genes,fill = Tissue))+
  geom_bar(stat = 'identity')+
  labs(x='', y = 'Tissue.Specific.Genes')+
  theme_classic()+
  theme(legend.position='none')+
  theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title =
          element_text(size=15))+
  theme(axis.text.x = element_text(angle = 90, vjust = .5, hjust = 1),
        panel.grid.major= element_blank(),panel.grid.minor = element_blank())+
  scale_fill_manual(values = accent_palette)
ggsave('Outputs/figures/DC2hm_tissue_enrichment_2.pdf', width = 6,
       height = 2.5)
ggplot(enrichmentOutput,aes(x=reorder(Tissue,Tissue.Specific.Genes),
                            y=Log10PValue,
                            label = Tissue.Specific.Genes,fill = Tissue))+
  geom_bar(stat = 'identity')+
  labs(x='', y = '-Log10PValue')+
  theme_bw()+
  theme(legend.position='none')+
  theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title =
          element_text(size=15))+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        panel.grid.major= element_blank(),panel.grid.minor = element_blank())+
  scale_fill_manual(values = set2_palette)+
  geom_hline(yintercept = 1.3, linetype='dashed', color='grey')
ggsave('Outputs/figures/DC2hm_tissue_enrichment_PVal_2.pdf', 
       width = 10, height = 4)
# Downregulated
gs <- GeneSet(
  geneIds =  intersect(dc2hm.down.2,
                       aire.induced.tra),
  organism = "Homo Sapiens",                 # Specify the organism
  geneIdType = SymbolIdentifier(),           # Define gene ID type as Symbol
  setName = "DC2hmGenes"               # Optional: Give a name to the gene set
)
dc2hm.tissue.enrich.dn.2 <- teEnrichment(gs)
# Fetch all hits
dc2hm.enrichment.hits.dn.2 <- c()
for (i in names(dc2hm.tissue.enrich.dn.2[[2]])) {
  dc2hm.enrichment.hits.dn.2 <- c(dc2hm.enrichment.hits.dn.2,
                                  rownames(dc2hm.tissue.enrich.dn.2[[2]][[i]]))
}
seEnrichmentOutput.dn <-dc2hm.tissue.enrich.dn.2[[1]]
enrichmentOutput.dn<-setNames(data.frame(assay(seEnrichmentOutput.dn),
                                         row.names = rowData(seEnrichmentOutput.dn)[,1]),
                              colData(seEnrichmentOutput)[,1])
enrichmentOutput.dn$Tissue<-row.names(enrichmentOutput.dn)
set2_palette <- colorRampPalette(brewer.pal(8, "Set2"))(35)
# Plotting the P-Values
ggplot(enrichmentOutput.dn,aes(x=reorder(Tissue,Tissue.Specific.Genes),
                               y=Tissue.Specific.Genes,
                               label = Tissue.Specific.Genes,fill = Tissue))+
  geom_bar(stat = 'identity')+
  labs(x='', y = 'Tissue.Specific.Genes')+
  theme_bw()+
  theme(legend.position='none')+
  theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title =
          element_text(size=15))+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        panel.grid.major= element_blank(),panel.grid.minor = element_blank())+
  scale_fill_manual(values = set2_palette)
ggsave('Outputs/figures/DC2hm_tissue_enrichment.dn_2.pdf', width = 10, height = 4)
ggplot(enrichmentOutput.dn,aes(x=reorder(Tissue,Tissue.Specific.Genes),
                               y=Log10PValue,
                               label = Tissue.Specific.Genes,fill = Tissue))+
  geom_bar(stat = 'identity')+
  labs(x='', y = '-Log10PValue')+
  theme_bw()+
  theme(legend.position='none')+
  theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title =
          element_text(size=15))+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        panel.grid.major= element_blank(),panel.grid.minor = element_blank())+
  scale_fill_manual(values = set2_palette)+
  geom_hline(yintercept = 1.3, linetype='dashed', color='grey')
ggsave('Outputs/figures/DC2hm_tissue_enrichment.dn_Pval_2.pdf', 
       width = 10, height = 4)
# Save TissueEnrich results
write.xlsx(list('DC2hm upregulated TRAs'=dc2hm.vs.cdc2.de.results[intersect(dc2hm.up.2, aire.induced.tra),],
                'DC2hm up TRA TissueEnrich '=enrichmentOutput,
                'DC2hm downregulated TRAs'=dc2hm.vs.cdc2.de.results[intersect(dc2hm.down.2, aire.induced.tra),],
                'DC2hm down TissueEnrich'=enrichmentOutput.dn
), 'Outputs/tables/DC2hm_TRA_analysis.xlsx', rowNames=T)
  # Show specific TRAs
combined.tra.show <- c("KCNN1", "NRXN2","GRIN1", "GRM5",
                       "ENTHD1", "POU4F1","ITIH3","PPP1R1C",
                       "COL7A1","PDGFRA")
dc2hm.vs.cdc2.de.results.show <- dc2hm.vs.cdc2.de.results[combined.tra.show ,]
dc2hm.vs.cdc2.de.results.show$Gene <- factor(rownames(dc2hm.vs.cdc2.de.results.show),
                                             levels = combined.tra.show )
ggplot(dc2hm.vs.cdc2.de.results.show,
       aes(x=Gene, y=logFC, fill=-log10(adj.P.Val)))+
  geom_bar(stat = 'identity')+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))+
  scale_fill_gradient(low = '#F686BD', high = '#d930a6')+
  ylab('DC2hm vs cDC2')
ggsave('Outputs/figures/Barplot_upregulated_TRAs.pdf', 
       width = 4, height = 2)
# Save the TRA and TissueEnrich results
write.xlsx(list('DC2hm upregulated TRAs'=dc2hm.vs.cdc2.de.results[intersect(dc2hm.up.2, aire.induced.tra),],
                'DC2hm up TRA TissueEnrich '=enrichmentOutput,
                'DC2hm downregulated TRAs'=dc2hm.vs.cdc2.de.results[intersect(dc2hm.down.2, aire.induced.tra),],
                'DC2hm down TissueEnrich'=enrichmentOutput.dn
), 'Outputs/tables/DC2hm_TRA_analysis.xlsx', rowNames=T)
